src/core/a-assets.js

```diff
@@ -10,126 +10,127 @@ var warn = debug('core:a-assets:warn');
-          return el.isAssetItem && el.hasAttribute('src');
-        });
+    // Wait for <audio>s and <video>s.
+    mediaEls = this.querySelectorAll('audio, video');
+    for (i = 0; i < mediaEls.length; i++) {
+      mediaEl = fixUpMediaElement(mediaEls[i]);
+      if (!mediaEl.src && !mediaEl.srcObject) {
+        warn('Audio/video asset has neither `src` nor `srcObject` attributes.');
       }
+      loaded.push(mediaElementLoaded(mediaEl));
     }
-  })
-});
+
+    // Trigger loaded for scene to start rendering.
+    Promise.all(loaded).then(bind(this.load, this));
+
+    // Timeout to start loading anyways.
+    timeout = parseInt(this.getAttribute('timeout'), 10) || 3000;
+    this.timeout = setTimeout(function () {
+      if (self.hasLoaded) { return; }
+      warn('Asset loading timed out in ', timeout, 'ms');
+      self.emit('timeout');
+      self.load();
+    }, timeout);
+  }
+
+  disconnectedCallback () {
+    // ANode method.
+    super.disconnectedCallback();
+
+    if (this.timeout) { clearTimeout(this.timeout); }
+  }
+
+  load () {
+    super.load.call(this, null, function waitOnFilter (el) {
+      return el.isAssetItem && el.hasAttribute('src');
+    });
+  }
+}
+
+customElements.define('a-assets', AAssets);
 
 /**
  * Preload using XHRLoader for any type of asset.
  */
-registerElement('a-asset-item', {
-  prototype: Object.create(ANode.prototype, {
-    createdCallback: {
-      value: function () {
-        this.data = null;
-        this.isAssetItem = true;
-      }
-    },
-
-    attachedCallback: {
-      value: function () {
-        var self = this;
-        var src = this.getAttribute('src');
-        fileLoader.setResponseType(
-          this.getAttribute('response-type') || inferResponseType(src));
-        fileLoader.load(src, function handleOnLoad (response) {
-          self.data = response;
-          /*
-            Workaround for a Chrome bug. If another XHR is sent to the same url before the
-            previous one closes, the second request never finishes.
-            setTimeout finishes the first request and lets the logic triggered by load open
-            subsequent requests.
-            setTimeout can be removed once the fix for the bug below ships:
-            https://bugs.chromium.org/p/chromium/issues/detail?id=633696&q=component%3ABlink%3ENetwork%3EXHR%20&colspec=ID%20Pri%20M%20Stars%20ReleaseBlock%20Component%20Status%20Owner%20Summary%20OS%20Modified
-          */
-          setTimeout(function load () { ANode.prototype.load.call(self); });
-        }, function handleOnProgress (xhr) {
-          self.emit('progress', {
-            loadedBytes: xhr.loaded,
-            totalBytes: xhr.total,
-            xhr: xhr
-          });
-        }, function handleOnError (xhr) {
-          self.emit('error', {xhr: xhr});
-        });
-      }
-    }
-  })
-});
+class AAssetItem extends ANode {
+  constructor (self) {
+    super(self);
```

Comment:
same as above, please consider adding an `init`, or `created` method to `ANode` class

```js
class Anode extends HTMLElement {
  constructor(...$) { const _ = super(...$); _.init(); return _; }
  init() { /* override as you like */ }}
}

class AAssetItem extends ANode {
  init() {
    this.data = null;
    this.isAssetItem = true;
  }
}
```

Reply:
I tried that but `init` would have to also remember to call `super.init` in derived classes.

Reply:
I guess my hint was rather for developers extending A-Frae components too, where writing `super.init();` instead of remembering the `self = super(self)` dance seems to be a win/more maintainable approach, but of course it's your choice.

The idea is that the day you won't need the polyfill anymore, you can simply rename all `init` as `constructor` and call it a day.